pkg/policy/repository.go

```diff
@@ -38,12 +41,28 @@ type Repository struct {
 	// incremented whenever the policy repository is changed.
 	// Always positive (>0).
 	revision uint64
+
+	// RepositoryChangeQueue is a queue which serializes changes to the policy
+	// repository.
+	RepositoryChangeQueue *eventqueue.EventQueue
+
+	// RuleReactionQueue is a queue which serializes the resultant events that
+	// need to occur after updating the state of the policy repository. This
+	// can include queueing endpoint regenerations, policy revision increments
+	// for endpoints, etc.
+	RuleReactionQueue *eventqueue.EventQueue
```

Comment:
These pieces seem like a really good candidate for a separate PR. Is there a strong motivation to keep them here? Was this reviewed in previous posts of this PR? (I've skipped these for now)

Reply:
The reason I had to add this synchronization was because synchronization must be performed between each API call which affects the state of the policy repository if I want to queue up events for each Endpoint that are dependent upon the nature of the policy change, which is the main functionality this PR introduces.

I can factor them out into another PR, but I have been asked in the past by various reviewers to keep changes which are necessary for other functionality in the same PR / commit.

Reply:
> The reason I had to add this synchronization was because synchronization must be performed between each API call which affects the state of the policy repository

Hmm. I had in my mind that the repo mutex today provided this, but then I realised that we've had this discussion already and to answer my question about "doesn't the repo mutex do this?", I can say no, it's not sufficient.

> I can factor them out into another PR, but I have been asked in the past by various reviewers to keep changes which are necessary for other functionality in the same PR / commit.

Right, having the example users is super useful during review. Here's where I'm coming from, I see two changes:
1. Change the policy ingestion handling from N:M goroutines of API handlers -> endpoint policy regenerators to N:1:M where there is this eventqueue in the middle that's responsible for serializing and passing on event requests. This is already desired as an independent change (see  for example #7220) 
1. Selectively regenerate endpoints (which includes needing to create policy revision bump events).

The (1) above seems like it could potentially introduce significant changes in policy handling behaviour because now we have a single new goroutine responsible for handling all policy events and distributing the required reactions to those events. New queues that may need to be tuned, there are potential changes to locking patterns from goroutines (eg on the repo lock, although at a glance it seems like that should improve with this change). From the perspective of changing one thing at a time and understanding the behaviour delta, it seems nicer to observe this change first.

That said, (2) is a super nice optimization and I agree (based on our discussions and re-reading the PR description) that (1) is required to implement (2). Furthermore, (2) will make drastic improvements to the amount of processing required to implement policy, which may improve the behaviour of the EventQueue in (1) simply by reducing the potential blocking time when queueing events to endpoints. So maybe from that perspective it doesn't matter so much to observe (1) in isolation because the average-case behaviour of the pipeline will drastically change with this anyway.

Summary answer to my question above: That part (1) can't be a follow-up PR, if it is split out then it would have to be merged first. Is it worth actually doing that? Not sure, I could be convinced either way at this point.